File: pipeline/adapters/libnova_adapter/src/pointing.c
#include "common.h"

/* ------------------------------------------------------------------ */
/* Experiment: equ_horizontal                                          */
/* Equatorial → Horizontal via libnova                                 */
/* libnova azimuth: 0°=South, increasing westward → convert to         */
/* 0°=North, increasing eastward to match ERFA convention.             */
/* Input per line: jd_ut1 jd_tt ra_rad dec_rad obs_lon_rad obs_lat_rad */
/* ------------------------------------------------------------------ */

void run_equ_horizontal(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"equ_horizontal\",\"library\":\"libnova\",");
    printf("\"model\":\"libnova_transform\",");
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_ut1, jd_tt, ra_rad, dec_rad, obs_lon, obs_lat;
        if (scanf("%lf %lf %lf %lf %lf %lf", &jd_ut1, &jd_tt, &ra_rad, &dec_rad,
                  &obs_lon, &obs_lat) != 6) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }

        double ra_deg  = ra_rad  * (180.0 / M_PI);
        double dec_deg = dec_rad * (180.0 / M_PI);
        double lon_deg = obs_lon * (180.0 / M_PI);
        double lat_deg = obs_lat * (180.0 / M_PI);

        struct ln_equ_posn equ = { .ra = ra_deg, .dec = dec_deg };
        struct ln_lnlat_posn observer = { .lng = lon_deg, .lat = lat_deg };
        struct ln_hrz_posn hrz;
        ln_get_hrz_from_equ(&equ, &observer, jd_ut1, &hrz);

        /* libnova: az 0=South, increasing toward West (CW from above).
         * ERFA:    az 0=North, increasing toward East (CW from above).
         * Convert: az_erfa = (az_libnova + 180°) mod 360° */
        double az_erfa_deg = fmod(hrz.az + 180.0, 360.0);
        double az_rad = az_erfa_deg * (M_PI / 180.0);
        double alt_rad = hrz.alt * (M_PI / 180.0);

        /* Closure via reverse */
        struct ln_hrz_posn hrz2 = { .az = hrz.az, .alt = hrz.alt };
        struct ln_equ_posn equ_back;
        ln_get_equ_from_hrz(&hrz2, &observer, jd_ut1, &equ_back);

        double ra_back  = equ_back.ra  * (M_PI / 180.0);
        double dec_back = equ_back.dec * (M_PI / 180.0);
        double v_in[3]  = { cos(dec_rad)*cos(ra_rad), cos(dec_rad)*sin(ra_rad), sin(dec_rad) };
        double v_bk[3]  = { cos(dec_back)*cos(ra_back), cos(dec_back)*sin(ra_back), sin(dec_back) };
        double closure_rad = ang_sep(v_in, v_bk);

        if (i > 0) printf(",\n");
        printf("{\"jd_ut1\":%.15f,\"jd_tt\":%.15f,", jd_ut1, jd_tt);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,", ra_rad, dec_rad);
        printf("\"obs_lon_rad\":%.17e,\"obs_lat_rad\":%.17e,", obs_lon, obs_lat);
        printf("\"az_rad\":%.17e,\"alt_rad\":%.17e,", az_rad, alt_rad);
        printf("\"closure_rad\":%.17e}", closure_rad);
    }
    printf("\n]}\n");
}

void run_equ_horizontal_perf(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    double *params = malloc(n * 6 * sizeof(double));

    for (int i = 0; i < n; i++) {
        if (scanf("%lf %lf %lf %lf %lf %lf",
                  &params[6*i], &params[6*i+1], &params[6*i+2],
                  &params[6*i+3], &params[6*i+4], &params[6*i+5]) != 6) {
            fprintf(stderr, "bad input line %d\n", i);
            exit(1);
        }
    }

    /* Pre-convert to libnova degree-based coordinates outside timing. */
    double *params_deg = malloc(n * 5 * sizeof(double));
    for (int i = 0; i < n; i++) {
        params_deg[5*i] = params[6*i];  /* jd_ut1 */
        params_deg[5*i+1] = params[6*i+2] * (180.0 / M_PI); /* ra */
        params_deg[5*i+2] = params[6*i+3] * (180.0 / M_PI); /* dec */
        params_deg[5*i+3] = params[6*i+4] * (180.0 / M_PI); /* lon */
        params_deg[5*i+4] = params[6*i+5] * (180.0 / M_PI); /* lat */
    }

    /* Warm-up */
    for (int i = 0, warmup = get_perf_warmup(); i < n && i < warmup; i++) {
        double jd_ut1 = params_deg[5*i];
        double ra = params_deg[5*i+1];
        double dec = params_deg[5*i+2];
        double lon = params_deg[5*i+3];
        double lat = params_deg[5*i+4];

        struct ln_equ_posn object = { ra, dec };
        struct ln_lnlat_posn observer = { lon, lat };
        struct ln_hrz_posn hrz;
        ln_get_hrz_from_equ(&object, &observer, jd_ut1, &hrz);
    }

    /* Timed run */
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        double jd_ut1 = params_deg[5*i];
        double ra = params_deg[5*i+1];
        double dec = params_deg[5*i+2];
        double lon = params_deg[5*i+3];
        double lat = params_deg[5*i+4];

        struct ln_equ_posn object = { ra, dec };
        struct ln_lnlat_posn observer = { lon, lat };
        struct ln_hrz_posn hrz;
        ln_get_hrz_from_equ(&object, &observer, jd_ut1, &hrz);
        sink += hrz.az;
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);
    double per_op_ns = elapsed_ns / n;

    printf("{\"experiment\":\"equ_horizontal_perf\",\"library\":\"libnova\",");
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, per_op_ns);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);

    free(params);
    free(params_deg);
}

/* ------------------------------------------------------------------ */
/* Experiment: horiz_to_equ                                            */
/* Horizontal → Equatorial via ln_get_equ_from_hrz                     */
/* Input: jd_ut1 jd_tt az_rad alt_rad obs_lon_rad obs_lat_rad         */
/* ------------------------------------------------------------------ */

void run_horiz_to_equ(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    printf("{\"experiment\":\"horiz_to_equ\",\"library\":\"libnova\",");
    printf("\"model\":\"libnova_transform\",");
    printf("\"count\":%d,\"cases\":[\n", n);

    for (int i = 0; i < n; i++) {
        double jd_ut1, jd_tt, az_rad, alt_rad, obs_lon, obs_lat;
        if (scanf("%lf %lf %lf %lf %lf %lf", &jd_ut1, &jd_tt, &az_rad, &alt_rad,
                  &obs_lon, &obs_lat) != 6) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }
        /* Convert ERFA convention (az: 0=N, CW) → libnova (az: 0=S, CW):
         * az_libnova = (az_erfa - 180) mod 360 */
        double az_deg_ln = fmod((az_rad * (180.0 / M_PI)) - 180.0 + 360.0, 360.0);
        double alt_deg = alt_rad * (180.0 / M_PI);
        double lon_deg = obs_lon * (180.0 / M_PI);
        double lat_deg = obs_lat * (180.0 / M_PI);

        struct ln_hrz_posn hrz = { .az = az_deg_ln, .alt = alt_deg };
        struct ln_lnlat_posn observer = { .lng = lon_deg, .lat = lat_deg };
        struct ln_equ_posn equ;
        ln_get_equ_from_hrz(&hrz, &observer, jd_ut1, &equ);

        double ra  = equ.ra  * (M_PI / 180.0);
        double dec = equ.dec * (M_PI / 180.0);

        /* Closure: equ → hrz → equ */
        struct ln_hrz_posn hrz2;
        ln_get_hrz_from_equ(&equ, &observer, jd_ut1, &hrz2);
        struct ln_equ_posn equ_back;
        ln_get_equ_from_hrz(&hrz2, &observer, jd_ut1, &equ_back);

        double ra_back  = equ_back.ra  * (M_PI / 180.0);
        double dec_back = equ_back.dec * (M_PI / 180.0);
        double v_in[3]  = { cos(dec)*cos(ra), cos(dec)*sin(ra), sin(dec) };
        double v_bk[3]  = { cos(dec_back)*cos(ra_back), cos(dec_back)*sin(ra_back), sin(dec_back) };
        double closure_rad = ang_sep(v_in, v_bk);

        if (i > 0) printf(",\n");
        printf("{\"jd_ut1\":%.15f,\"jd_tt\":%.15f,", jd_ut1, jd_tt);
        printf("\"az_rad\":%.17e,\"alt_rad\":%.17e,", az_rad, alt_rad);
        printf("\"obs_lon_rad\":%.17e,\"obs_lat_rad\":%.17e,", obs_lon, obs_lat);
        printf("\"ra_rad\":%.17e,\"dec_rad\":%.17e,", ra, dec);
        printf("\"closure_rad\":%.17e}", closure_rad);
    }
    printf("\n]}\n");
}

void run_horiz_to_equ_perf(void) {
    int n;
    if (scanf("%d", &n) != 1) { fprintf(stderr, "bad N\n"); exit(1); }

    double *params = malloc(n * 6 * sizeof(double));
    for (int i = 0; i < n; i++) {
        if (scanf("%lf %lf %lf %lf %lf %lf",
                  &params[6*i], &params[6*i+1], &params[6*i+2],
                  &params[6*i+3], &params[6*i+4], &params[6*i+5]) != 6) {
            fprintf(stderr, "bad input line %d\n", i); exit(1);
        }
    }

    /* Pre-convert to libnova degree-based coordinates */
    double *p_deg = malloc(n * 5 * sizeof(double));
    for (int i = 0; i < n; i++) {
        p_deg[5*i]   = params[6*i]; /* jd_ut1 */
        double az_deg_ln = fmod((params[6*i+2] * (180.0 / M_PI)) - 180.0 + 360.0, 360.0);
        p_deg[5*i+1] = az_deg_ln;
        p_deg[5*i+2] = params[6*i+3] * (180.0 / M_PI); /* alt */
        p_deg[5*i+3] = params[6*i+4] * (180.0 / M_PI); /* lon */
        p_deg[5*i+4] = params[6*i+5] * (180.0 / M_PI); /* lat */
    }

    for (int i = 0, warmup = get_perf_warmup(); i < n && i < warmup; i++) {
        struct ln_hrz_posn hrz = { p_deg[5*i+1], p_deg[5*i+2] };
        struct ln_lnlat_posn obs = { p_deg[5*i+3], p_deg[5*i+4] };
        struct ln_equ_posn equ;
        ln_get_equ_from_hrz(&hrz, &obs, p_deg[5*i], &equ);
    }

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    double sink = 0.0;
    for (int i = 0; i < n; i++) {
        struct ln_hrz_posn hrz = { p_deg[5*i+1], p_deg[5*i+2] };
        struct ln_lnlat_posn obs = { p_deg[5*i+3], p_deg[5*i+4] };
        struct ln_equ_posn equ;
        ln_get_equ_from_hrz(&hrz, &obs, p_deg[5*i], &equ);
        sink += equ.ra + equ.dec;
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ns = (t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec);

    printf("{\"experiment\":\"horiz_to_equ_perf\",\"library\":\"libnova\",");
    printf("\"count\":%d,\"total_ns\":%.0f,\"per_op_ns\":%.1f,", n, elapsed_ns, elapsed_ns / n);
    printf("\"throughput_ops_s\":%.0f,\"_sink\":%.17e}\n",
           (double)n / (elapsed_ns * 1e-9), sink);
    free(params); free(p_deg);
}

File: pipeline/adapters/siderust_adapter/src/pointing.rs
use std::io::{self, Write};
use std::time::Instant;

use crate::affn;
use crate::qtty::{Degrees, Meters, Radians};

use siderust::coordinates::cartesian;
use siderust::coordinates::centers::Geodetic;
use siderust::coordinates::frames::{EquatorialTrueOfDate, Horizontal, ECEF};
use siderust::coordinates::transform::horizontal::FromHorizontal;
use siderust::coordinates::transform::DirectionAstroExt;
use siderust::time::JulianDate;

use crate::ang_sep;

pub(crate) fn run_equ_horizontal(lines: &mut impl Iterator<Item = String>) {
    let n: usize = lines.next().unwrap().trim().parse().unwrap();
    let stdout = io::stdout();
    let mut out = stdout.lock();

    writeln!(
        out,
        "{{\"experiment\":\"equ_horizontal\",\"library\":\"siderust\",\
         \"model\":\"siderust_horizontal\",\
         \"count\":{},\"cases\":[",
        n
    )
    .unwrap();

    for i in 0..n {
        let line = lines.next().unwrap();
        let parts: Vec<f64> = line
            .split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect();
        let jd_ut1 = parts[0];
        let jd_tt_val = parts[1];
        let ra_rad = parts[2];
        let dec_rad = parts[3];
        let obs_lon_rad = parts[4];
        let obs_lat_rad = parts[5];

        let jd_ut1_q = JulianDate::new(jd_ut1);
        let jd_tt_q = JulianDate::new(jd_tt_val);

        // Create equatorial true-of-date cartesian direction
        let spherical_equ = affn::spherical::Direction::<EquatorialTrueOfDate>::new_raw(
            Degrees::new(dec_rad.to_degrees()),
            Degrees::new(ra_rad.to_degrees()),
        );
        let equatorial_direction = spherical_equ.to_cartesian();

        // Create observer site
        let site = Geodetic::<ECEF>::new(
            Degrees::new(obs_lon_rad.to_degrees()),
            Degrees::new(obs_lat_rad.to_degrees()),
            Meters::new(0.0),
        );

        // Transform to horizontal (precise: explicit UT1+TT for benchmark fairness)
        let horizontal_direction = DirectionAstroExt::to_horizontal_precise(
            &equatorial_direction,
            &jd_tt_q,
            &jd_ut1_q,
            &site,
        );
        let horizontal_spherical = horizontal_direction.to_spherical();
        let az = Radians::from(horizontal_spherical.azimuth);
        let alt = Radians::from(horizontal_spherical.polar);

        // Transform back to equatorial
        let equatorial_back: affn::cartesian::Direction<EquatorialTrueOfDate> =
            horizontal_direction.to_equatorial(&jd_ut1_q, &jd_tt_q, &site);
        let equatorial_back_spherical = equatorial_back.to_spherical();
        let ra_back = Radians::from(equatorial_back_spherical.azimuth);
        let dec_back = Radians::from(equatorial_back_spherical.polar);

        let v_in = [
            dec_rad.cos() * ra_rad.cos(),
            dec_rad.cos() * ra_rad.sin(),
            dec_rad.sin(),
        ];
        let v_back = [
            dec_back.value().cos() * ra_back.value().cos(),
            dec_back.value().cos() * ra_back.value().sin(),
            dec_back.value().sin(),
        ];
        let closure_rad = ang_sep(&v_in, &v_back);

        if i > 0 {
            writeln!(out, ",").unwrap();
        }
        write!(
            out,
            "{{\"jd_ut1\":{:.15},\"jd_tt\":{:.15},\
             \"ra_rad\":{:.17e},\"dec_rad\":{:.17e},\
             \"obs_lon_rad\":{:.17e},\"obs_lat_rad\":{:.17e},\
             \"az_rad\":{:.17e},\"alt_rad\":{:.17e},\
             \"closure_rad\":{:.17e}}}",
            jd_ut1,
            jd_tt_val,
            ra_rad,
            dec_rad,
            obs_lon_rad,
            obs_lat_rad,
            az.value(),
            alt.value(),
            closure_rad,
        )
        .unwrap();
    }
    writeln!(out, "\n]}}").unwrap();
}

pub(crate) fn run_equ_horizontal_perf(lines: &mut impl Iterator<Item = String>) {
    let n: usize = lines.next().unwrap().trim().parse().unwrap();

    let mut params: Vec<(f64, f64, f64, f64, f64, f64)> = Vec::with_capacity(n);

    for _ in 0..n {
        let line = lines.next().unwrap();
        let parts: Vec<f64> = line
            .split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect();
        params.push((parts[0], parts[1], parts[2], parts[3], parts[4], parts[5]));
    }

    // Warm-up
    let warmup = crate::perf_warmup();
    for &(jd_ut1, jd_tt, ra, dec, lon, lat) in params.iter().take(n.min(warmup)) {
        let jd_ut1_q = JulianDate::new(jd_ut1);
        let jd_tt_q = JulianDate::new(jd_tt);

        let spherical_equ = affn::spherical::Direction::<EquatorialTrueOfDate>::new_raw(
            Degrees::new(dec.to_degrees()),
            Degrees::new(ra.to_degrees()),
        );
        let equatorial_direction = spherical_equ.to_cartesian();
        let site = Geodetic::<ECEF>::new(
            Degrees::new(lon.to_degrees()),
            Degrees::new(lat.to_degrees()),
            Meters::new(0.0),
        );
        let dir_hor = equatorial_direction.to_horizontal_precise(&jd_tt_q, &jd_ut1_q, &site);
        std::hint::black_box(&dir_hor);
    }

    // Timed run
    let start = Instant::now();
    let mut sink: (f64, f64) = (0.0, 0.0);
    for &(jd_ut1, jd_tt, ra, dec, lon, lat) in &params {
        let jd_ut1_q = JulianDate::new(jd_ut1);
        let jd_tt_q = JulianDate::new(jd_tt);

        let spherical_equ = affn::spherical::Direction::<EquatorialTrueOfDate>::new_raw(
            Degrees::new(dec.to_degrees()),
            Degrees::new(ra.to_degrees()),
        );
        let equatorial_direction = spherical_equ.to_cartesian();
        let site = Geodetic::<ECEF>::new(
            Degrees::new(lon.to_degrees()),
            Degrees::new(lat.to_degrees()),
            Meters::new(0.0),
        );
        let horizontal_direction =
            equatorial_direction.to_horizontal_precise(&jd_tt_q, &jd_ut1_q, &site);
        let horizontal_spherical = horizontal_direction.to_spherical();
        let az = Radians::from(horizontal_spherical.azimuth);
        let alt = Radians::from(horizontal_spherical.polar);
        sink = (az.value(), alt.value());
    }
    let elapsed = start.elapsed();
    let total_ns = elapsed.as_nanos() as f64;
    let per_op_ns = total_ns / n as f64;

    println!(
        "{{\"experiment\":\"equ_horizontal_perf\",\"library\":\"siderust\",\
         \"count\":{},\"total_ns\":{:.0},\"per_op_ns\":{:.1},\
         \"throughput_ops_s\":{:.0},\"_sink\":{:.17e}}}",
        n,
        total_ns,
        per_op_ns,
        n as f64 / (total_ns * 1e-9),
        sink.0,
    );
}

pub fn run_horiz_to_equ(lines: &mut impl Iterator<Item = String>) {
    let n: usize = lines.next().unwrap().trim().parse().unwrap();
    let stdout = io::stdout();
    let mut out = stdout.lock();

    writeln!(
        out,
        "{{\"experiment\":\"horiz_to_equ\",\"library\":\"siderust\",\
         \"model\":\"siderust_horizontal_inverse\",\
         \"count\":{},\"cases\":[",
        n
    )
    .unwrap();

    for i in 0..n {
        let line = lines.next().unwrap();
        let parts: Vec<f64> = line
            .split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect();
        let jd_ut1_val = parts[0];
        let jd_tt_val = parts[1];
        let az_rad = parts[2];
        let alt_rad = parts[3];
        let obs_lon_rad = parts[4];
        let obs_lat_rad = parts[5];

        let jd_ut1 = JulianDate::new(jd_ut1_val);
        let jd_tt = JulianDate::new(jd_tt_val);
        let site = Geodetic::<ECEF>::new(
            Degrees::new(obs_lon_rad.to_degrees()),
            Degrees::new(obs_lat_rad.to_degrees()),
            Meters::new(0.0),
        );

        let sph_hor = affn::spherical::Direction::<Horizontal>::new_raw(
            Degrees::new(alt_rad.to_degrees()),
            Degrees::new(az_rad.to_degrees()),
        );
        let hor_dir = sph_hor.to_cartesian();

        let equ_dir: cartesian::Direction<EquatorialTrueOfDate> =
            hor_dir.to_equatorial(&jd_ut1, &jd_tt, &site);
        let equ_sph = equ_dir.to_spherical();
        let ra = Radians::from(equ_sph.azimuth);
        let dec = Radians::from(equ_sph.polar);

        let hor_back = DirectionAstroExt::to_horizontal_precise(&equ_dir, &jd_tt, &jd_ut1, &site);
        let equ_back: cartesian::Direction<EquatorialTrueOfDate> =
            hor_back.to_equatorial(&jd_ut1, &jd_tt, &site);
        let equ_back_sph = equ_back.to_spherical();
        let ra_back = Radians::from(equ_back_sph.azimuth);
        let dec_back = Radians::from(equ_back_sph.polar);

        let v_in = [
            dec.value().cos() * ra.value().cos(),
            dec.value().cos() * ra.value().sin(),
            dec.value().sin(),
        ];
        let v_back = [
            dec_back.value().cos() * ra_back.value().cos(),
            dec_back.value().cos() * ra_back.value().sin(),
            dec_back.value().sin(),
        ];
        let closure_rad = ang_sep(&v_in, &v_back);

        if i > 0 {
            writeln!(out, ",").unwrap();
        }
        write!(
            out,
            "{{\"jd_ut1\":{:.15},\"jd_tt\":{:.15},\
             \"az_rad\":{:.17e},\"alt_rad\":{:.17e},\
             \"obs_lon_rad\":{:.17e},\"obs_lat_rad\":{:.17e},\
             \"ra_rad\":{:.17e},\"dec_rad\":{:.17e},\
             \"closure_rad\":{:.17e}}}",
            jd_ut1_val,
            jd_tt_val,
            az_rad,
            alt_rad,
            obs_lon_rad,
            obs_lat_rad,
            ra.value(),
            dec.value(),
            closure_rad,
        )
        .unwrap();
    }
    writeln!(out, "\n]}}").unwrap();
}

pub fn run_horiz_to_equ_perf(lines: &mut impl Iterator<Item = String>) {
    let n: usize = lines.next().unwrap().trim().parse().unwrap();

    let mut params: Vec<(f64, f64, f64, f64, f64, f64)> = Vec::with_capacity(n);
    for _ in 0..n {
        let line = lines.next().unwrap();
        let parts: Vec<f64> = line
            .split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect();
        params.push((parts[0], parts[1], parts[2], parts[3], parts[4], parts[5]));
    }

    let warmup = crate::perf_warmup();
    for &(jd_ut1_v, jd_tt_v, az, alt, lon, lat) in params.iter().take(n.min(warmup)) {
        let jd_ut1 = JulianDate::new(jd_ut1_v);
        let jd_tt = JulianDate::new(jd_tt_v);
        let site = Geodetic::<ECEF>::new(
            Degrees::new(lon.to_degrees()),
            Degrees::new(lat.to_degrees()),
            Meters::new(0.0),
        );
        let sph = affn::spherical::Direction::<Horizontal>::new_raw(
            Degrees::new(alt.to_degrees()),
            Degrees::new(az.to_degrees()),
        );
        let hor = sph.to_cartesian();
        let d: cartesian::Direction<EquatorialTrueOfDate> =
            hor.to_equatorial(&jd_ut1, &jd_tt, &site);
        std::hint::black_box(&d);
    }

    let start = Instant::now();
    let mut sink: f64 = 0.0;
    for &(jd_ut1_v, jd_tt_v, az, alt, lon, lat) in &params {
        let jd_ut1 = JulianDate::new(jd_ut1_v);
        let jd_tt = JulianDate::new(jd_tt_v);
        let site = Geodetic::<ECEF>::new(
            Degrees::new(lon.to_degrees()),
            Degrees::new(lat.to_degrees()),
            Meters::new(0.0),
        );
        let sph = affn::spherical::Direction::<Horizontal>::new_raw(
            Degrees::new(alt.to_degrees()),
            Degrees::new(az.to_degrees()),
        );
        let hor = sph.to_cartesian();
        let equ: cartesian::Direction<EquatorialTrueOfDate> =
            hor.to_equatorial(&jd_ut1, &jd_tt, &site);
        let equ_sph = equ.to_spherical();
        let ra = Radians::from(equ_sph.azimuth);
        let dec = Radians::from(equ_sph.polar);
        sink += ra.value() + dec.value();
    }
    let elapsed = start.elapsed();
    let total_ns = elapsed.as_nanos() as f64;

    println!(
        "{{\"experiment\":\"horiz_to_equ_perf\",\"library\":\"siderust\",\
         \"count\":{},\"total_ns\":{:.0},\"per_op_ns\":{:.1},\
         \"throughput_ops_s\":{:.0},\"_sink\":{:.17e}}}",
        n,
        total_ns,
        total_ns / n as f64,
        n as f64 / (total_ns * 1e-9),
        sink,
    );
}
